### PR TITLE
gstd documentation

### PR DESCRIPTION
**Pull request by AndrePanin**
Please review, provide your edits if necessary (by meaning), after I will proofread it.

### CODE REVIEW

File: gstd/src/exec.rs [hunk 1]
```diff
@@ -16,19 +16,123 @@
...
+/// and put into the *processing queue*.
+///
+/// # Examples
+///
+/// ```
+/// use gstd::{exec, MessageId};
+///
+/// pub unsafe extern "C" fn handle() {
+///     // ...
+///     exec::wake(MessageId::default());
```
Comment (breathx) [inline gstd/src/exec.rs:115]:
```suggestion
///     ...
///     let msg_id = ... ;
///     exec::wake(msg_id);
```

Found it more correct

File: gstd/src/lib.rs [hunk 2]
```diff
@@ -16,6 +16,10 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
+/// Declares modules, attributes, public re-exports.
+/// All Gear crates don't load Rust std modules (no_std) which makes them lightweight.
```
Comment (breathx) [inline gstd/src/lib.rs:20]:
```suggestion
/// Gear libs are `#![no_std]`, which makes them lightweight. 
```

File: gstd/src/lib.rs [hunk 3]
```diff
@@ -16,6 +16,10 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
+/// Declares modules, attributes, public re-exports.
+/// All Gear crates don't load Rust std modules (no_std) which makes them lightweight. 
+/// They are based on the core library.
```
Comment (breathx) [inline gstd/src/lib.rs:21]:
```suggestion
```

I think this is superfluous information

File: gstd/src/lock/mutex.rs [hunk 4]
```diff
@@ -16,7 +16,11 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Mutex async implementation.
+//! Classic async implementation of mutex.
```
Comment (breathx) [inline gstd/src/lock/mutex.rs:19]:
```suggestion
//! Mutex async implementation.
```

File: gstd/src/lock/mutex.rs [hunk 5]
```diff
@@ -97,6 +103,8 @@ pub struct MutexLockFuture<'a, T> {
     mutex: &'a Mutex<T>,
 }
 
+/// In case of locked mutex and an `.await`, function `poll` checks if the mutex can be taken, else it waits.
```
Comment (breathx) [inline gstd/src/lock/mutex.rs:106]:
```suggestion
/// In case of locked mutex and an `.await`, function `poll` checks if the mutex can be taken, else it waits (goes into *waiting queue*).
```

File: gstd/src/macros/metadata.rs [hunk 6]
```diff
@@ -16,7 +16,11 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Gear `metadata!` macro. Exports functions with IO data.
+//! Gear `metadata!` macro serves for exporting functions from Rust with 
+//! IO data that the user has specified in the macro.
+//! Macro contains `input` and `output` message types for `init`, `handle`,
+//! `async`. It is used in `idea.gear-tech.io` in order to correctly form
```
Comment (breathx) [inline gstd/src/macros/metadata.rs:22]: It also contains `state` IO for reading state and reading it, depending on some arguments passed

File: gstd/src/msg/async.rs [hunk 7]
```diff
@@ -16,7 +16,8 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Module with async messaging functions.
+//! This `gstd` module provides async messaging functions. 
+//! While `gcore` provides API for low-level async implementation only (gcore::exec).
```
Comment (breathx) [inline gstd/src/msg/async.rs:20]:
```suggestion
```

I thinks that it's superfluous too

File: gstd/src/msg/async.rs [hunk 8]
```diff
@@ -61,6 +62,15 @@ impl<D: Decode> FusedFuture for CodecMessageFuture<D> {
     }
 }
 
+/// To interrupt a program execition waiting for a reply on a previous message,
```
Comment (breathx) [inline gstd/src/msg/async.rs:65]:
```suggestion
/// To interrupt a program execution waiting for a reply on a previous message,
```

File: gstd/src/msg/async.rs [hunk 9]
```diff
@@ -61,6 +62,15 @@ impl<D: Decode> FusedFuture for CodecMessageFuture<D> {
...
+/// one need to specify an `.await` expression.
```
Comment (breathx) [inline gstd/src/msg/async.rs:66]: probably `to call .await` would be more suitable

File: gstd/src/msg/async.rs [hunk 10]
```diff
@@ -61,6 +62,15 @@ impl<D: Decode> FusedFuture for CodecMessageFuture<D> {
...
+/// The initial message that requires a reply is sent instantly.
+/// methods `send_and_wait_for_reply` retun `MessageFuture` where `.await` can be specified.
```
Comment (breathx) [inline gstd/src/msg/async.rs:68]:
```suggestion
/// methods `send_and_wait_for_reply` return `MessageFuture` which implements `Future` trait.
```

File: gstd/src/msg/async.rs [hunk 11]
```diff
@@ -61,6 +62,15 @@ impl<D: Decode> FusedFuture for CodecMessageFuture<D> {
...
+/// `MessageFuture` checks if the reply exists already:
+///     if not, program interrupts
+///     if reply exists, function returns the payload `Result`
+///     if reply came without paylod, program  with `ContractError::ExitCode`
```
Comment (breathx) [inline gstd/src/msg/async.rs:72]:
```suggestion
///     if reply exists, function checks it's exit code:
///          if it equals 0, returns `Ok(payload)`
///          else returns `Err(ContractError::ExitCode(code))`
```
```suggestion
///     if reply exists, function returns the payload `Result`
///     if reply came without paylod, program  with `ContractError::ExitCode`
```

note: Anyway need to fmt it

File: gstd/src/msg/async.rs [hunk 12]
```diff
@@ -82,15 +92,15 @@ impl Future for MessageFuture {
...
+} 
 
 impl FusedFuture for MessageFuture {
     fn is_terminated(&self) -> bool {
         !signals().waits_for(self.waiting_reply_to)
     }
 }
 
-/// Send a message and wait for reply.
+/// Send a message and wait for reply (as encoded structure).
```
Comment (breathx) [inline gstd/src/msg/async.rs:103]: Need to somehow say that input should be codec Encodable, while output is codec Decodable

File: gstd/src/msg/encoded.rs [hunk 13]
```diff
@@ -16,7 +16,16 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Module with encoded via scale-codec messaging functions.
+//! Module with messaging functions (load, reply, send) for operating 
+//! with messages arguments as with data structure instead of bytes array 
+//! decoded/encoded via SCALE Codec.
```
Comment (breathx) [inline gstd/src/msg/encoded.rs:21]: Should we ref parity-scale-codec lib?

File: gstd/src/msg/encoded.rs [hunk 14]
```diff
@@ -16,7 +16,16 @@
...
+/// `load` returns Result of decoding because of possible errors.
+/// If ok, it returns the structure itself, else error.
```
Comment (breathx) [inline gstd/src/msg/encoded.rs:23]:
```suggestion
```

it returns `Result` structure

File: gstd/src/msg/encoded.rs [hunk 15]
```diff
@@ -16,7 +16,16 @@
...
+//! 
+//! Example:
+//! ```
+//! let x: String = load().expect(msg: &str);
```
Comment (breathx) [inline gstd/src/msg/encoded.rs:27]:
```suggestion
//! use gstd::msg;
//! ...
//! let x: String = msg::load().expect("Unable to decode `String`");
```

File: gstd/src/prelude.rs [hunk 16]
```diff
@@ -16,7 +16,8 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! The `gstd` default prelude.
+//! The `gstd` default prelude. Imports `gstd` modules and traits into each program.
```
Comment (breathx) [inline gstd/src/prelude.rs:19]:
```suggestion
//! The `gstd` default prelude. Re-imports default `std` modules and traits.
```

File: gstd/src/msg/basic.rs [hunk 17]
```diff
@@ -63,36 +101,227 @@ impl From<gcore::MessageHandle> for MessageHandle {
...
+/// obtained using the [`size`] function.
+///
+/// # Examples
+///
+/// ```
+/// use gstd::msg;
+///
+/// pub unsafe extern "C" fn handle() {
+///     let mut result = vec![0u8; msg::size()];
+///     msg::load_bytes(&mut result[..]);
```
Comment (breathx) [inline gstd/src/msg/basic.rs:155]:
```suggestion
///     let payload_bytes = msg::load_bytes();
```

File: gstd/src/msg/basic.rs [hunk 18]
```diff
@@ -63,36 +101,227 @@ impl From<gcore::MessageHandle> for MessageHandle {
...
+/// use gstd::{msg, ActorId};
+///
+/// pub unsafe extern "C" fn handle() {
+///     // ...
+///     let mut id: [u8; 32] = [0; 32];
+///     for i in 0..id.len() {
+///         id[i] = i as u8;
+///     }
+///
+///     msg::send_bytes(ActorId(id), b"HELLO", 1000, 12345678);
```
Comment (breathx) [inline gstd/src/msg/basic.rs:316]:
```suggestion
///     let id = msg::source();
///
///     msg::send_bytes(ActorId::new(id), b"HELLO", 1000, 12345678);
```

**Review by breathx — CHANGES_REQUESTED** [review]
Need to fix some issues, I've mentioned above.

Probably, need to wait other's comments and re-check formulations.

Speaking about the informational payload, looks good to me.

File: gstd/src/common/primitives.rs [hunk 19]
```diff
@@ -25,6 +40,15 @@ use codec::{Decode, Encode};
 use primitive_types::H256;
 use scale_info::TypeInfo;
 
+/// Program (actor) identifier.
+///
+/// Gear allows users and programs to interact with other users and programs via
+/// messages. Source and target program as well as user are represented by
+/// 256-bit identifier `ActorId` struct. The source `ActorId` for a message
+/// being processed can be obtained using [`msg::source`](crate::msg::source)
+/// function. Also, each send function has a target `ActorId` as one of the
```
[the remark this answers is not in the document]
Comment (AndrePanin) [reply gstd/src/common/primitives.rs:49]: send and reply are different functions, so it looks like correct.

File: gstd/src/macros/metadata.rs (hunk 6 above)
Comment (AndrePanin) [reply gstd/src/macros/metadata.rs:22]: added

File: gstd/src/msg/async.rs (hunk 9 above)
Comment (AndrePanin) [reply gstd/src/msg/async.rs:65]: ok

File: gstd/src/msg/encoded.rs (hunk 13 above)
Comment (AndrePanin) [reply gstd/src/msg/encoded.rs:21]: I suppose no, it is in the 1st place in google results.

File: gstd/src/msg/async.rs [hunk 20]
```diff
@@ -61,6 +61,15 @@ impl<D: Decode> FusedFuture for CodecMessageFuture<D> {
     }
 }
 
+/// To interrupt a program execution waiting for a reply on a previous message,
+/// one need to call an `.await` expression.
+/// The initial message that requires a reply is sent instantly.
+/// Methods `send_and_wait_for_reply` return `MessageFuture` where `.await` can
```
Comment (peter-cmw) [inline gstd/src/msg/async.rs:67]: @breathx please check if description is ok

File: gstd/src/msg/async.rs [hunk 21]
```diff
@@ -90,7 +99,7 @@ impl FusedFuture for MessageFuture {
     }
 }
 
-/// Send a message and wait for reply.
+/// Send a message and wait for reply (as encoded structure).
```
Comment (peter-cmw) [inline gstd/src/msg/async.rs:102]: I would rephrase a bit - specify that wait for reply and returns result as a structure versus same thing with raw reply message payload

File: gstd/src/msg/encoded.rs [hunk 22]
```diff
@@ -16,7 +16,17 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Module with encoded via scale-codec messaging functions.
+//! Module with messaging functions (load, reply, send) for operating
+//! with messages arguments as with data structure instead of bytes array
+//! decoded/encoded via SCALE Codec.
+//! `load` returns Result of decoding because of possible errors.
```
Comment (peter-cmw) [inline gstd/src/msg/encoded.rs:22]: ?? because of possible errors?

File: gstd/src/prelude.rs [hunk 23]
```diff
@@ -16,7 +16,9 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! The `gstd` default prelude.
+//! The `gstd` default prelude. Re-imports default `std` modules and traits.
+//! It is safe for Rust programs to rename `std` to `gstd` so that nothing
```
Comment (peter-cmw) [inline gstd/src/prelude.rs:20]: I would rephrase this

**Review by peter-cmw — CHANGES_REQUESTED** [review]

File: gstd/src/macros/metadata.rs [hunk 24]
```diff
@@ -16,7 +16,15 @@
...
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-//! Gear `metadata!` macro. Exports functions with IO data.
+//! Gear `metadata!` macro serves for exporting functions from Rust with
+//! IO data that the user has specified in the macro.
+//! Macro contains `input` and `output` message types for `init`, `handle`,
+//! `async`. It also contains `state` IO for reading state and reading it,
+//! depending on some arguments passed.
+//! Metadata can be used as a payload description for external tools and
+//! applications that work with the blockchain network.
```
[the remark this answers is not in the document]
Comment (AndrePanin) [reply gstd/src/macros/metadata.rs:25]: ok

File: gstd/src/macros/metadata.rs [hunk 25]
```diff
@@ -16,7 +16,15 @@
...
+//! For example, it is used in `idea.gear-tech.io` in order to correctly
+//! form a payload from json on the JS application side.
```
[the remark this answers is not in the document]
Comment (AndrePanin) [reply gstd/src/macros/metadata.rs:27]: ok

File: gstd/src/msg/encoded.rs (hunk 22 above)
Comment (AndrePanin) [reply gstd/src/msg/encoded.rs:22]: rephrased

File: gstd/src/prelude.rs (hunk 23 above)
Comment (AndrePanin) [reply gstd/src/prelude.rs:20]: ok

File: gstd/codegen/src/lib.rs [hunk 26]
```diff
@@ -29,12 +29,15 @@ fn compile_error<T: ToTokens>(tokens: T, msg: &str) -> TokenStream {
         .into()
 }
 
-/// Mark async function to be the program entry point.
+/// This is the procedural macro for your convenience.
+/// It marks main async function to be the program entry point.
+/// Functions handle, handle_reply canont be specified if this macro is used.
```
Comment (shamilsan) [inline gstd/codegen/src/lib.rs:34]:
```suggestion
/// Functions handle, handle_reply cannot be specified if this macro is used.
```

File: gstd/src/msg/async.rs (hunk 20 above)
Comment (breathx) [reply gstd/src/msg/async.rs:67]: I've already commented on this description. I still think that formulation like "return `MessageFuture` which implements `Future` trait" is more correct neither than given "where `.await` can be called", because Rust's async is common and understandable thing

Probably it will be more clearly if we describe with words return types instead of directly point at `Result` variants, but I doubt is it better for documentation.

I also suppose that "if reply was received" is a bit complete explanation than "if reply exists"

Comment (AndrePanin) [reply gstd/src/msg/async.rs]: updated

**Review by peter-cmw — APPROVED** [review]

**Review by breathx — APPROVED** [review]
